src/responder/nss/nsssrv_mmap_cache.c

```diff
@@ -548,17 +548,25 @@ static struct sss_mc_rec *sss_mc_find_record(struct sss_mc_ctx *mcc,
         }
 
         safealign_memcpy(&name_ptr, rec->data, sizeof(rel_ptr_t), NULL);
-        if (key->len > strs_len
-            || (name_ptr + key->len) > (strs_offset + strs_len)
-            || (uint8_t *)rec->data + strs_offset + strs_len > max_addr) {
+        t_key = (char *)rec->data + name_ptr;
+        /* name_ptr must point to some data in the strs/gids area of the data
+         * payload. Since it is a pointer relative to rec->data it must larger
+         * equal strs_offset and must be smaller then strs_offset + strs_len.
+         * Additionally the area must not end outside of the data table and
+         * t_key must be a zero-terminates string. */
+        if (name_ptr < strs_offset
+                || name_ptr >= strs_offset + strs_len
+                || (uint8_t *)rec->data + strs_offset + strs_len > max_addr
+                || memchr(t_key, '\0',
+                          (strs_offset + strs_len) - name_ptr) == NULL ) {
```

Comment:
All these checks are there because we assume that memory cache can be corrupted(mmap cache consist of blocks with binary data and with NUL terminated strings)

And in such case we cannot trust that `strs_len`. Moreover if mmap cache is corrupted and and t_key points to one of last data blocks + following data blocks does not contain character `'\0'` then memchr can easily go out of bounds.

Reply:
Also here we can add a check that e.g. strs_len is not larger then mcc->dt_size. But even if strs_len is corrupted the '(uint8_t *)rec->data + strs_offset + strs_len > max_addr' check should prevent that memchr() goes out of bounds.

Reply:
Ok, the check might overrun. @lslebodn, do you think it is ok to use GCC's builtin __builtin_add_overflow() or trying to prevent the overrun by other means?

Reply:
> But even if strs_len is corrupted the '(uint8_t *)rec->data + strs_offset + strs_len > max_addr' 

Ahh that's true but in such case we can directly try `strcmp`. Because it will be within bounds of data table and input(`key`) is NUL terminated => usage of `strcmp` is safe even with corrupted/binary data in mmap cache.